### DISCUSSION TITLE
Building an application for both Windows and Linux

### DISCUSSION THREAD

**Discussion by alexpolmans** [airsdk/Adobe-Runtime-Support#1852] [General]
I have an application that runs under Windows with Flex and a captive runtime.   Now I want to port it to Linux, again with a captive runtime, and am trying to understand the process I must follow.
1. Do I have to make the Linux Build on a Linux Platform using the Linux SDK for Flex developers?
2. Can I download the Windows SDK for Flex Developers, and build on a Windows Machine
3. Can have a single build on a Windows Machine and then add the runtime at the end (and how would I do that)?
Any pointers would be welcome.

**Comment by alexpolmans** [comment]
I have some progress here to share ...

I installed Windows Subsystem for Linux to get a Linux development environment that can share source files with Windows
I compile a small test program 
`.../sdks/Flex 4.16.1 with AIR 33.1.1 Linux/bin/amxmlc -load-config .../sdks/Flex 4.16.1 with AIR 33.1.1 Linux/frameworks/air-config.xml -output bin-release-linux/TestLinux.swf -headless-server=true src/TestLinux.mxml`
This gives me the required swf file, which I build using
`.../sdks/Flex 4.16.1 with AIR 33.1.1 Linux/bin/adt -package -tsa http://sha256timestamp.ws.symantec.com/sha256/timestamp -storetype pkcs12 -keystore XYZ.pfx -storepass XXXXX -target native TestLinux.air TestLinux-app.xml TestLinux.swf`

This command fails with the following message
`Could not generate timestamp: sha256timestamp.ws.symantec.com`

Has anyone else tried building a Linux target under WSL?

**Reply by ajwfrost** [reply]
Hi

We should probably just get ADT to bomb out when it encounters `-target native` on Linux... this option is used to create a native installer that does not include the AIR runtime, i.e. when it's run, it would need the shared AIR runtime to be already present on the machine. And we don't have a shared AIR runtime for Linux.

So it would be better to use `-target bundle` which creates a folder that includes the application along with the captive AIR runtime, and then turn this into an installation package via other Linux tools (or just zip it up)..

thanks

**Comment by GeneralVimes** [comment]
I have experience of an AIR game on Windows (without Flex). I work on Windows machine. To make a Linux build I compiled .swf on Windows, then moved it together with the application.xml and assets file to a machine with Ubuntu and successfully compiled a Linux application there. Described my steps here: https://www.reddit.com/r/as3/comments/p9vqd2/building_air_game_for_linux_working_experience/

Again, it’s not about flex, but maybe something can be helpful

**Comment by ajwfrost** [comment]
Just responding to the last of your questions in the original post ... currently when you create a bundle for each platform, it would need to be run on the platform that you're using i.e. to create a linux bundle, you run the packager on linux, to create a windows bundle, you run it from a windows machine, etc...

Some of the processes that happen may be machine-specific i.e. relying on other tools that are part of Windows/Mac. For Linux .. I don't think that this is actually the case. So it might be possible for you to create a bundle for Linux, just using the AIR SDK (for Linux) from a Windows or Mac machine. But you would need to manually go through the steps that the packager would normally do, and if there are ANEs involved then this is probably a bit convoluted.

I had wondered about the idea of 'cross compiling' to create a Linux binary from a Windows machine etc, the challenge being that the way the AIR runtime is built, the appropriate tools and processes are built into this based on the source files applicable to the target platform... so the Windows runtime contains the capability to package apps for Windows, and trying to get it to also contain the capability to package apps for Linux would require a redesign of the whole process and inheritance/OO concepts...

Hopefully though, WSL and WSA are going to be quite useful tools for Windows users to be able to package, build and run stuff for Linux/Android whilst not leaving the Windows environment... and we're looking at a way to package up the AIR SDK so that you can have multiple variants coexisting i.e. currently if you download the Linux AIR SDK there's a `runtimes/air/linux` folder, and if you download the Windows AIR SDK then there's a `runtimes/air/win` folder... so we can have a generic SDK that has both of those at the same time.

thanks

**Reply by alexpolmans** [reply]
Thank you both.
I've spent the morning installing VirtualBox, and getting it configured to access the windows file system, and install the java development kit (JDK)   Without any other real changes I was able to get ADT to build and saw the same directory structure described by GeneralVimes.

So JVM is *not* a suitable environment if you need to timestamp the build.

Now the issue is running the program.   At this point all, I have is a small flex program, which I have attached to this port. (remove the .txt at the end of the name).

It fails with the following message.

`_Gtk-Message: 11:48:33.344: Failed to load module "canberra-gtk-module"`

Is there something else I should be including?




![image](https://user-images.githubusercontent.com/38087730/169859195-d0f2a2d9-738c-4c7a-a383-9cb7d715b633.png)
[TestLinux.mxml.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/8756161/TestLinux.mxml.txt)
[TestLinux-app.xml.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/8756162/TestLinux-app.xml.txt)

**Comment by alexpolmans** [comment]
Hi Andrew,

I have a small test program compiling and building and when I try to run it I seem to be blocked from accessing the Linux Runtime.

***@***.***

Is this covered by our existing license?    Developer Id = fbb267e3-fa88-41f3-be56-b4208d213ddd under ***@***.******@***.***>

If it is, how do I get past this;    if not, how can we get access?

Thanks

Alex

**Reply by ajwfrost** [reply]
Hi

Github is playing around with the message a little here .. but can I check what you're actually seeing when you refer to this:
> I seem to be blocked from accessing the Linux Runtime

If you run the app from the command-line/terminal window, is there anything being reported there?

Are you just using ADL at this point, or is this for a bundle? I did notice one problem with our bundle template in one of the recent SDKs where the file permissions weren't set up properly, so you may need to run `chmod a+x AppBinaryName` where this is the binary filename in the generated folder (i.e. if you run `file AppBinaryName` it should tell you it's an ELF 64-bit executable file...)

The running of AIR on Linux require a commercial license to be present - for ADL, this means within the SDK, but when you package an app with ADT you should end up with a license.txt file alongside your application descriptor. So as long as that's present and correct, the launch should be okay..

thanks

**Comment by alexpolmans** [comment]
Thanks Andrew,

I am running from the command line, which is where I see the following:

***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$ ./TestLinux
*******************************************************************************
Adobe AIR - Linux runtime is limited to use by commercially licensed developers
*******************************************************************************
***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$ ls
'Adobe AIR'   META-INF   mimetype   TestLinux   TestLinux.swf
***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$


and there is no license.txt.

I have a meeting now; I’ll search for it afterwards.

Thanks

Alex

**Comment by alexpolmans** [comment]
Sorry – to answer your question – I’m using ADT to build the file; roughly

    bash "$ADT" \
        -package \
        -tsa http://sha256timestamp.ws.symantec.com/sha256/timestamp \
        -storetype pkcs12 \
        -keystore "$CERT_PATH" \
        -storepass @@@@ \
        -target bundle \
        "$APPLICATION_NAME" \
        "$APPLICATION_NAME-app.xml" \
        "$APPLICATION_NAME.swf"




Thanks Andrew,

I am running from the command line, which is where I see the following:

***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$ ./TestLinux
*******************************************************************************
Adobe AIR - Linux runtime is limited to use by commercially licensed developers
*******************************************************************************
***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$ ls
'Adobe AIR'   META-INF   mimetype   TestLinux   TestLinux.swf
***@***.***:/media/sf_alex.polmans/git/N2KView Workspace/n2kview/bin-release-linux/TestLinux$


and there is no license.txt.

I have a meeting now; I’ll search for it afterwards.

Thanks

Alex

**Reply by ajwfrost** [reply]
Okay thanks

So I would suggest:
- can you see what happens when you run `adt -license` on the command line there, to check it knows that there's a license?
- the adt.lic file is in transition currently, we should be coping with this in multiple locations, but it may be worth duplicating it: either in the AIR SDK "lib" folder, or in `~/.airsdk/` 
- if you use ADT to package up a Linux bundle whilst it's aware of the license file, then a license.txt file should be in the generated bundle under the `META-INF/AIR/` folder

thanks

**Comment by alexpolmans** [comment]
Hi Andrew

bash: /media/sf_alex.polmans/git/N2KView Workspace/n2kcommon/sdks/Flex 4.16.1 with AIR 33.1.1 Linux/bin/adt -license: No such file or directory

Interestingly I cannot find a copy of our license.txt file on my computer, and I have record of receiving it after we last renewed our license in October last year.  Is there a chance that you could resend it to me? (Shows how long its been since I did a build 😊

Alex

**Reply by ajwfrost** [reply]
Hi

Your error message suggests you just need to put quotes around the command, i.e.
```
"/media/sf_alex.polmans/git/N2KView Workspace/n2kcommon/sdks/Flex 4.16.1 with AIR 33.1.1 Linux/bin/adt" -license
```
since the `No such file or directory` I think is saying that it doesn't recognise the requested path.

Plus -> within the AIR SDK, it's a file called `adt.lic` that it will be looking for; the `license.txt` is an application-specific file that's generated when you package up an application with ADT.

But for sure, I can trigger a re-sending of the license file, let me go look up our details for doing that....

thanks

**Comment by alexpolmans** [comment]
Thanks Andrew,

I may only get back to this tomorrow;  Ill let you know how it goes.

Alex

**Comment by alexpolmans** [comment]
Hi Andrew,

IT WORKED

Adding the license file back in was the last step.

And a slight adjustment of the quotes in the ADT -license shows

AIR Developer Tool version 33.1.1.821
License file is present and valid
DeveloperID=fbb267e3-fa88-41f3-be56-b4208d213ddd

Thank you so much for your help.

Regards

Alex

**Reply by ajwfrost** [reply]
Great, thanks for confrming.
